{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ba0e9669",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_csv from `anndata` is deprecated. Import anndata.io.read_csv instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_excel from `anndata` is deprecated. Import anndata.io.read_excel instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_hdf from `anndata` is deprecated. Import anndata.io.read_hdf instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_loom from `anndata` is deprecated. Import anndata.io.read_loom instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_mtx from `anndata` is deprecated. Import anndata.io.read_mtx instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_text from `anndata` is deprecated. Import anndata.io.read_text instead.\n",
      "  warnings.warn(msg, FutureWarning)\n",
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/anndata/utils.py:434: FutureWarning: Importing read_umi_tools from `anndata` is deprecated. Import anndata.io.read_umi_tools instead.\n",
      "  warnings.warn(msg, FutureWarning)\n"
     ]
    }
   ],
   "source": [
    "import sys, importlib\n",
    "import scanpy as sc\n",
    "import os\n",
    "import anndata as ad\n",
    "from sklearn.model_selection import train_test_split\n",
    "import numpy as np\n",
    "\n",
    "sys.path.append(\"..\")\n",
    "import scripts.data_prep as dp\n",
    "importlib.reload(dp)\n",
    "\n",
    "IN_PATH_neu = \"../data/processed/GSE169569_GSE169570_neutrophils_bowel_disease-25.07.0.h5ad\"\n",
    "IN_PATH_cov = \"../data/processed/GSE228841_PRJNA951718_COVID-19_vaccine-25.07.0.h5ad\"\n",
    "\n",
    "adata_neu = sc.read_h5ad(IN_PATH_neu)\n",
    "adata_cov = sc.read_h5ad(IN_PATH_cov)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ff97091",
   "metadata": {},
   "source": [
    "### Check information of the datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "51969496",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['total_counts_before_preprocessing', 'total_counts_after_trimming',\n",
      "       'total_counts_after_preprocessing',\n",
      "       'total_count_ratio__after_to_before', 'QC_mapping_ratio_bacterial',\n",
      "       'QC_mapping_ratio_viral', 'QC_mapping_ratio_miRNA',\n",
      "       'QC_mapping_ratio_sncRNA', 'Bases', 'Bytes', 'Avg_spot_length',\n",
      "       'BioProject', 'BioSample', 'Experiment', 'GEO_Accession_exp',\n",
      "       'SRA_study', 'Sample_name', 'Consent', 'Center_name', 'Organism',\n",
      "       'source_name', 'cohort', 'sex', 'age', 'treatment', 'diagnosis',\n",
      "       'Assay_type', 'Sequencer', 'Sample_type', 'Lab_library_layout',\n",
      "       'Lab_library_selection', 'Lab_library_source',\n",
      "       'Lab_RNA_extraction_protocol', 'Lab_Blocking_State',\n",
      "       'Lab_library_preparation_kit', 'Release_date', 'create_date', 'version',\n",
      "       'total_count'],\n",
      "      dtype='object')\n",
      "------------------    ------------------  ------------------    ------------------\n",
      "Index(['total_counts_before_preprocessing', 'total_counts_after_preprocessing',\n",
      "       'total_count_ratio__after_to_before', 'QC_mapping_ratio_bacterial',\n",
      "       'QC_mapping_ratio_viral', 'QC_mapping_ratio_miRNA',\n",
      "       'QC_mapping_ratio_sncRNA', 'Bases', 'Bytes', 'Avg_spot_length',\n",
      "       'BioProject', 'BioSample', 'Experiment', 'SRA_study', 'Sample_name',\n",
      "       'Consent', 'Center_name', 'Organism', 'source_name',\n",
      "       'Participant_number', 'sex', 'age', 'vaccine', 'naat_result',\n",
      "       'timepoint', 'Assay_type', 'Sequencer', 'Sample_type',\n",
      "       'Lab_library_layout', 'Lab_library_selection', 'Lab_library_source',\n",
      "       'Lab_RNA_extraction_protocol', 'Lab_Blocking_State',\n",
      "       'Lab_library_preparation_kit', 'Release_date', 'create_date', 'version',\n",
      "       'total_count'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "print(adata_neu.obs.columns)\n",
    "print(\"------------------    ------------------  ------------------    ------------------\")\n",
    "print(adata_cov.obs.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4c14f253",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['length', 'g_fraction', 'a_fraction', 't_fraction', 'c_fraction',\n",
      "       'gc_fraction', 'MFE_37', 'spikein', 'qiaseq_spikein', 'hbdx_spikein',\n",
      "       ...\n",
      "       'MS2_Cancer__Thrombocytes__RPMscaled_prop',\n",
      "       'MS2_Cancer__Monocytes__RPMscaled_prop',\n",
      "       'MS2_Cancer__B_cells__RPMscaled_prop',\n",
      "       'MS2_Cancer__Basophils__RPMscaled_prop',\n",
      "       'MS2_Cancer__Eosinophils__RPMscaled_prop',\n",
      "       'MS2_Cancer__Erythrocytes__RPMscaled_prop',\n",
      "       'MS2_Cancer__main_cell_type__100', 'MS2_Cancer__main_cell_type__75',\n",
      "       'MS2_Cancer__main_cell_type__50', 'MS2_Cancer__plasma_cell_RPM_ratio'],\n",
      "      dtype='object', length=186)\n",
      "------------------    ------------------  ------------------    ------------------\n",
      "Index(['length', 'g_fraction', 'a_fraction', 't_fraction', 'c_fraction',\n",
      "       'gc_fraction', 'MFE_37', 'spikein', 'qiaseq_spikein', 'hbdx_spikein',\n",
      "       ...\n",
      "       'MS2_Cancer__Thrombocytes__RPMscaled_prop',\n",
      "       'MS2_Cancer__Monocytes__RPMscaled_prop',\n",
      "       'MS2_Cancer__B_cells__RPMscaled_prop',\n",
      "       'MS2_Cancer__Basophils__RPMscaled_prop',\n",
      "       'MS2_Cancer__Eosinophils__RPMscaled_prop',\n",
      "       'MS2_Cancer__Erythrocytes__RPMscaled_prop',\n",
      "       'MS2_Cancer__main_cell_type__100', 'MS2_Cancer__main_cell_type__75',\n",
      "       'MS2_Cancer__main_cell_type__50', 'MS2_Cancer__plasma_cell_RPM_ratio'],\n",
      "      dtype='object', length=186)\n"
     ]
    }
   ],
   "source": [
    "print(adata_neu.var.columns)\n",
    "print(\"------------------    ------------------  ------------------    ------------------\")\n",
    "print(adata_cov.var.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "dd32000d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['GTGCATGATCTCAAGTTTTCAATCTGAGACCT', 'ATCACAGGGTAGAACCACGGAC',\n",
       "       'TGGAGAGAAAGGCAGTTCCTGT', 'TCCTGACTCCAGGTCCTGTGT',\n",
       "       'CGTGGTCTCCCAACCCTTGTACCAGT', 'TCGAGGACCCCCCCTGCCTGG',\n",
       "       'TCGGGCCGATCGCACGCC', 'CTGGGAATACCGGGTGCTGTAGGCTA',\n",
       "       'TGACCTATGAATTGGCAGCC', 'AGGTTCCGGATAAGTAAGAGCT',\n",
       "       ...\n",
       "       'TCTCGTCTGATCTCGGAAGCTAAGCAGGGTCGGG', 'AATCCCGGACGAGCCCTG',\n",
       "       'ATTCGTAGACGACCTGCTTCTGGGTCGGGGTT', 'GAAAGATGGTGAACTATGCCTGGGCAGGG',\n",
       "       'TTCAAGTAATCCAGGATAGGCTT', 'TTCCGTACTGAGCTGCCCCGA',\n",
       "       'CTGGCCCTCTCTGCCCTTAAGA', 'AACCGAGCGTCCAAGCTCTTT',\n",
       "       'TTTTCATTAATCAAGAACGAAAGTCGGAGG',\n",
       "       'GAGACCCGTCGCCGCGCTCTCCCCCCTCCCGGCGCC'],\n",
       "      dtype='object', name='Sequence', length=51777)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_neu.var_names               # Index([...])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c790e957",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>BioProject</th>\n",
       "      <th>Center_name</th>\n",
       "      <th>Organism</th>\n",
       "      <th>source_name</th>\n",
       "      <th>cohort</th>\n",
       "      <th>sex</th>\n",
       "      <th>age</th>\n",
       "      <th>treatment</th>\n",
       "      <th>diagnosis</th>\n",
       "      <th>Assay_type</th>\n",
       "      <th>Sequencer</th>\n",
       "      <th>Sample_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>smpID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>SRR14062693</th>\n",
       "      <td>PRJNA717018</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>peripheral blood</td>\n",
       "      <td>Swedish</td>\n",
       "      <td>male</td>\n",
       "      <td>27</td>\n",
       "      <td>treatment-naive</td>\n",
       "      <td>Symptomatic control</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAXgene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR14062694</th>\n",
       "      <td>PRJNA717018</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>peripheral blood</td>\n",
       "      <td>Swedish</td>\n",
       "      <td>female</td>\n",
       "      <td>20</td>\n",
       "      <td>treatment-naive</td>\n",
       "      <td>Ulcerative colitis</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAXgene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR14062695</th>\n",
       "      <td>PRJNA717018</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>peripheral blood</td>\n",
       "      <td>Swedish</td>\n",
       "      <td>female</td>\n",
       "      <td>30</td>\n",
       "      <td>treatment-naive</td>\n",
       "      <td>Symptomatic control</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAXgene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR14062696</th>\n",
       "      <td>PRJNA717018</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>peripheral blood</td>\n",
       "      <td>Swedish</td>\n",
       "      <td>male</td>\n",
       "      <td>27</td>\n",
       "      <td>treatment-naive</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAXgene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR14062697</th>\n",
       "      <td>PRJNA717018</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>peripheral blood</td>\n",
       "      <td>Swedish</td>\n",
       "      <td>male</td>\n",
       "      <td>32</td>\n",
       "      <td>treatment-naive</td>\n",
       "      <td>Ulcerative colitis</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAXgene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR14063135</th>\n",
       "      <td>PRJNA717025</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>peripheral blood</td>\n",
       "      <td>German</td>\n",
       "      <td>male</td>\n",
       "      <td>28</td>\n",
       "      <td>treatment-exposed</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAXgene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR14063136</th>\n",
       "      <td>PRJNA717025</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>peripheral blood</td>\n",
       "      <td>German</td>\n",
       "      <td>female</td>\n",
       "      <td>66</td>\n",
       "      <td>treatment-exposed</td>\n",
       "      <td>Ulcerative colitis</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAXgene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR14063137</th>\n",
       "      <td>PRJNA717025</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>peripheral blood</td>\n",
       "      <td>German</td>\n",
       "      <td>male</td>\n",
       "      <td>23</td>\n",
       "      <td>treatment-exposed</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAXgene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR14063138</th>\n",
       "      <td>PRJNA717025</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>peripheral blood</td>\n",
       "      <td>German</td>\n",
       "      <td>female</td>\n",
       "      <td>43</td>\n",
       "      <td>treatment-exposed</td>\n",
       "      <td>Crohn's disease</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAXgene</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SRR14063139</th>\n",
       "      <td>PRJNA717025</td>\n",
       "      <td>GEO</td>\n",
       "      <td>Homo sapiens</td>\n",
       "      <td>peripheral blood</td>\n",
       "      <td>German</td>\n",
       "      <td>female</td>\n",
       "      <td>43</td>\n",
       "      <td>treatment-exposed</td>\n",
       "      <td>Ulcerative colitis</td>\n",
       "      <td>ncRNA-Seq</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>PAXgene</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>447 rows Ã— 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              BioProject Center_name      Organism       source_name   cohort  \\\n",
       "smpID                                                                           \n",
       "SRR14062693  PRJNA717018         GEO  Homo sapiens  peripheral blood  Swedish   \n",
       "SRR14062694  PRJNA717018         GEO  Homo sapiens  peripheral blood  Swedish   \n",
       "SRR14062695  PRJNA717018         GEO  Homo sapiens  peripheral blood  Swedish   \n",
       "SRR14062696  PRJNA717018         GEO  Homo sapiens  peripheral blood  Swedish   \n",
       "SRR14062697  PRJNA717018         GEO  Homo sapiens  peripheral blood  Swedish   \n",
       "...                  ...         ...           ...               ...      ...   \n",
       "SRR14063135  PRJNA717025         GEO  Homo sapiens  peripheral blood   German   \n",
       "SRR14063136  PRJNA717025         GEO  Homo sapiens  peripheral blood   German   \n",
       "SRR14063137  PRJNA717025         GEO  Homo sapiens  peripheral blood   German   \n",
       "SRR14063138  PRJNA717025         GEO  Homo sapiens  peripheral blood   German   \n",
       "SRR14063139  PRJNA717025         GEO  Homo sapiens  peripheral blood   German   \n",
       "\n",
       "                sex  age          treatment            diagnosis Assay_type  \\\n",
       "smpID                                                                         \n",
       "SRR14062693    male   27    treatment-naive  Symptomatic control  ncRNA-Seq   \n",
       "SRR14062694  female   20    treatment-naive   Ulcerative colitis  ncRNA-Seq   \n",
       "SRR14062695  female   30    treatment-naive  Symptomatic control  ncRNA-Seq   \n",
       "SRR14062696    male   27    treatment-naive      Crohn's disease  ncRNA-Seq   \n",
       "SRR14062697    male   32    treatment-naive   Ulcerative colitis  ncRNA-Seq   \n",
       "...             ...  ...                ...                  ...        ...   \n",
       "SRR14063135    male   28  treatment-exposed      Crohn's disease  ncRNA-Seq   \n",
       "SRR14063136  female   66  treatment-exposed   Ulcerative colitis  ncRNA-Seq   \n",
       "SRR14063137    male   23  treatment-exposed      Crohn's disease  ncRNA-Seq   \n",
       "SRR14063138  female   43  treatment-exposed      Crohn's disease  ncRNA-Seq   \n",
       "SRR14063139  female   43  treatment-exposed   Ulcerative colitis  ncRNA-Seq   \n",
       "\n",
       "                       Sequencer Sample_type  \n",
       "smpID                                         \n",
       "SRR14062693  Illumina HiSeq 2500     PAXgene  \n",
       "SRR14062694  Illumina HiSeq 2500     PAXgene  \n",
       "SRR14062695  Illumina HiSeq 2500     PAXgene  \n",
       "SRR14062696  Illumina HiSeq 2500     PAXgene  \n",
       "SRR14062697  Illumina HiSeq 2500     PAXgene  \n",
       "...                          ...         ...  \n",
       "SRR14063135  Illumina HiSeq 2500     PAXgene  \n",
       "SRR14063136  Illumina HiSeq 2500     PAXgene  \n",
       "SRR14063137  Illumina HiSeq 2500     PAXgene  \n",
       "SRR14063138  Illumina HiSeq 2500     PAXgene  \n",
       "SRR14063139  Illumina HiSeq 2500     PAXgene  \n",
       "\n",
       "[447 rows x 12 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# batch for neu: bioproject, center_name, sequencer, lab_library_preparation_kit/lab_rna_extraction_protocol\n",
    "# batch for cov: bioproject, SRA_study, center_name, sequencer, lab_* fields\n",
    "adata_neu.obs[['BioProject','Center_name', 'Organism', 'source_name', 'cohort', 'sex', 'age', 'treatment', 'diagnosis', 'Assay_type', 'Sequencer','Sample_type']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a08e557",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: (447, 51777)\n",
      "X dtype: float64\n",
      "Example values from X: [0.         0.         0.65439486 0.         0.         0.\n",
      " 0.65439486 0.81989226 0.         6.50348223]\n",
      "Layers: []\n",
      "\n",
      ".obs columns: ['total_counts_before_preprocessing', 'total_counts_after_trimming', 'total_counts_after_preprocessing', 'total_count_ratio__after_to_before', 'QC_mapping_ratio_bacterial', 'QC_mapping_ratio_viral', 'QC_mapping_ratio_miRNA', 'QC_mapping_ratio_sncRNA', 'Bases', 'Bytes', 'Avg_spot_length', 'BioProject', 'BioSample', 'Experiment', 'GEO_Accession_exp', 'SRA_study', 'Sample_name', 'Consent', 'Center_name', 'Organism', 'source_name', 'cohort', 'sex', 'age', 'treatment', 'diagnosis', 'Assay_type', 'Sequencer', 'Sample_type', 'Lab_library_layout', 'Lab_library_selection', 'Lab_library_source', 'Lab_RNA_extraction_protocol', 'Lab_Blocking_State', 'Lab_library_preparation_kit', 'Release_date', 'create_date', 'version', 'total_count']\n",
      "\n",
      ".obs sample (first 5 rows):\n",
      "             total_counts_before_preprocessing  total_counts_after_trimming  \\\n",
      "smpID                                                                         \n",
      "SRR14062693                            6288582                      5833092   \n",
      "SRR14062694                            8410144                      7842618   \n",
      "SRR14062695                            6709769                      6206453   \n",
      "SRR14062696                            8254214                      7682426   \n",
      "SRR14062697                            6099701                      5695046   \n",
      "\n",
      "             total_counts_after_preprocessing  \\\n",
      "smpID                                           \n",
      "SRR14062693                           5240117   \n",
      "SRR14062694                           7065181   \n",
      "SRR14062695                           5585572   \n",
      "SRR14062696                           6965692   \n",
      "SRR14062697                           5159983   \n",
      "\n",
      "             total_count_ratio__after_to_before  QC_mapping_ratio_bacterial  \\\n",
      "smpID                                                                         \n",
      "SRR14062693                              0.8332                      0.0001   \n",
      "SRR14062694                              0.8400                      0.0001   \n",
      "SRR14062695                              0.8324                      0.0001   \n",
      "SRR14062696                              0.8438                      0.0000   \n",
      "SRR14062697                              0.8459                      0.0001   \n",
      "\n",
      "             QC_mapping_ratio_viral  QC_mapping_ratio_miRNA  \\\n",
      "smpID                                                         \n",
      "SRR14062693                       0                  0.9276   \n",
      "SRR14062694                       0                  0.9319   \n",
      "SRR14062695                       0                  0.9362   \n",
      "SRR14062696                       0                  0.9185   \n",
      "SRR14062697                       0                  0.9442   \n",
      "\n",
      "             QC_mapping_ratio_sncRNA      Bases      Bytes  ...  \\\n",
      "smpID                                                       ...   \n",
      "SRR14062693                   0.0315  320717682  188674814  ...   \n",
      "SRR14062694                   0.0366  428917344  244956956  ...   \n",
      "SRR14062695                   0.0310  342198219  201134186  ...   \n",
      "SRR14062696                   0.0486  420964914  238779206  ...   \n",
      "SRR14062697                   0.0229  311084751  179988237  ...   \n",
      "\n",
      "             Lab_library_layout Lab_library_selection Lab_library_source  \\\n",
      "smpID                                                                      \n",
      "SRR14062693              single    size fractionation     transcriptomic   \n",
      "SRR14062694              single    size fractionation     transcriptomic   \n",
      "SRR14062695              single    size fractionation     transcriptomic   \n",
      "SRR14062696              single    size fractionation     transcriptomic   \n",
      "SRR14062697              single    size fractionation     transcriptomic   \n",
      "\n",
      "                Lab_RNA_extraction_protocol Lab_Blocking_State  \\\n",
      "smpID                                                            \n",
      "SRR14062693  PAXgene Blood RNA Kit (Qiagen)        no blocking   \n",
      "SRR14062694  PAXgene Blood RNA Kit (Qiagen)        no blocking   \n",
      "SRR14062695  PAXgene Blood RNA Kit (Qiagen)        no blocking   \n",
      "SRR14062696  PAXgene Blood RNA Kit (Qiagen)        no blocking   \n",
      "SRR14062697  PAXgene Blood RNA Kit (Qiagen)        no blocking   \n",
      "\n",
      "            Lab_library_preparation_kit          Release_date  \\\n",
      "smpID                                                           \n",
      "SRR14062693   Illumina TruSeq Small RNA  2022-08-17T00:00:00Z   \n",
      "SRR14062694   Illumina TruSeq Small RNA  2022-08-17T00:00:00Z   \n",
      "SRR14062695   Illumina TruSeq Small RNA  2022-08-17T00:00:00Z   \n",
      "SRR14062696   Illumina TruSeq Small RNA  2022-08-17T00:00:00Z   \n",
      "SRR14062697   Illumina TruSeq Small RNA  2022-08-17T00:00:00Z   \n",
      "\n",
      "                      create_date version total_count  \n",
      "smpID                                                  \n",
      "SRR14062693  2021-03-24T21:30:00Z       1     5226885  \n",
      "SRR14062694  2021-03-24T21:31:00Z       1     7047728  \n",
      "SRR14062695  2021-03-24T21:30:00Z       1     5570886  \n",
      "SRR14062696  2021-03-24T21:31:00Z       1     6951350  \n",
      "SRR14062697  2021-03-24T21:30:00Z       1     5149287  \n",
      "\n",
      "[5 rows x 39 columns]\n",
      "\n",
      ".var columns: ['length', 'g_fraction', 'a_fraction', 't_fraction', 'c_fraction', 'gc_fraction', 'MFE_37', 'spikein', 'qiaseq_spikein', 'hbdx_spikein', 'hbdx_spikein_prefix_length', 'hbdx_spikein_suffix_length', 'hbdx_spikein_affix_filter', 'five_prime_adapter_length', 'five_prime_adapter_filter', 'adapter_mapping_filter', 'kmer_five_prime_adapter', 'kmer_three_prime_adapter', 'kmer_filter', 'housekeeping', 'TE_annotation', 'bacterial', 'viral', 'blocking_deseq2R_log2FoldChange', 'blocking_deseq2R_padj', 'reduced_by_blocking', 'blocking_oligo__duplex_energy', 'intended_blocking_target', 'chr', 'num_genomic_loci_maps', 'cluster_gene_name', 'tumor_prog_markers', 'tumor_prog_markers_mutualinfo', 'tumor_prog_markers_meantrend', 'tcga_LC_DESeq2_bool', 'tcga_LC_DESeq2', 'tcga_LC_DESeq2_log2FC', 'tcga_LC_DESeq2_padj', 'tcga_LUSC_DESeq2_bool', 'tcga_LUAD_DESeq2_bool', 'tcga_normal__highest_tissue', 'tcga_normal__highest_mean', 'tcga_normal__TSI', 'tcga_cancer__LUAD_LUSC_DE', 'tcga_cancer__DE_cancers_log2FC', 'tcga_cancer__nr_nonLC_DE_cancers', 'Sozzi', 'Montani', 'HBDx_Masterplate', 'HBDx_IP', 'JAMA', 'General_Literature', 'single_class_annotation', 'single_name_annotation', 'small_RNA_class_annotation', 'subclass_name', 'subclass_name_new_binning', 'subclass_name_new_binning_pos', 'subclass_type', 'pseudo_class', 'mitochondrial', 'ambiguity_resolved', 'hico', 'mms', 'reference', 'precursor_name_full', 'ref_miR_seq', 'miRNA_ref_iso', 'miRNA_hico', 'miRNA_seed', 'miRNA_names', 'tRNA_ref_iso', 'MirGeneDB_ID', 'PTM_count', 'PTM_NGS_mean', 'protein_count', 'protein_affinity_median', 'protein_coverage_perc_median', 'RBCs_RPM_mean', 'CD56+_RPM_mean', 'Neutrophile_RPM_mean', 'CD4+_RPM_mean', 'CD19+_RPM_mean', 'Platelets_RPM_mean', 'Eosinophile_RPM_mean', 'CD8+_RPM_mean', 'CD14+_RPM_mean', 'Basophile_RPM_mean', 'RBCs_RPM_std', 'CD56+_RPM_std', 'Neutrophile_RPM_std', 'CD4+_RPM_std', 'CD19+_RPM_std', 'Platelets_RPM_std', 'Eosinophile_RPM_std', 'CD8+_RPM_std', 'CD14+_RPM_std', 'Basophile_RPM_std', 'RBCs_RPM_prop', 'CD56+_RPM_prop', 'Neutrophile_RPM_prop', 'CD4+_RPM_prop', 'CD19+_RPM_prop', 'Platelets_RPM_prop', 'Eosinophile_RPM_prop', 'CD8+_RPM_prop', 'CD14+_RPM_prop', 'Basophile_RPM_prop', 'Platelets_RPMscaled_mean', 'CD14+_RPMscaled_mean', 'CD56+_RPMscaled_mean', 'CD8+_RPMscaled_mean', 'RBCs_RPMscaled_mean', 'Neutrophile_RPMscaled_mean', 'CD19+_RPMscaled_mean', 'CD4+_RPMscaled_mean', 'Eosinophile_RPMscaled_mean', 'Basophile_RPMscaled_mean', 'Platelets_RPMscaled_std', 'CD14+_RPMscaled_std', 'CD56+_RPMscaled_std', 'CD8+_RPMscaled_std', 'RBCs_RPMscaled_std', 'Neutrophile_RPMscaled_std', 'CD19+_RPMscaled_std', 'CD4+_RPMscaled_std', 'Eosinophile_RPMscaled_std', 'Basophile_RPMscaled_std', 'Platelets_RPMscaled_prop', 'CD14+_RPMscaled_prop', 'CD56+_RPMscaled_prop', 'CD8+_RPMscaled_prop', 'RBCs_RPMscaled_prop', 'Neutrophile_RPMscaled_prop', 'CD19+_RPMscaled_prop', 'CD4+_RPMscaled_prop', 'Eosinophile_RPMscaled_prop', 'Basophile_RPMscaled_prop', 'main_cell_type__100', 'main_cell_type__75', 'main_cell_type__50', 'MS2_all__Plasma_Norgen__RPMscaled_prop', 'MS2_all__Neutrophils__RPMscaled_prop', 'MS2_all__CD8__RPMscaled_prop', 'MS2_all__NK_cells__RPMscaled_prop', 'MS2_all__CD4__RPMscaled_prop', 'MS2_all__Monocytes__RPMscaled_prop', 'MS2_all__B_cells__RPMscaled_prop', 'MS2_all__Erythrocytes__RPMscaled_prop', 'MS2_all__Thrombocytes__RPMscaled_prop', 'MS2_all__Basophils__RPMscaled_prop', 'MS2_all__Eosinophils__RPMscaled_prop', 'MS2_all__main_cell_type__100', 'MS2_all__main_cell_type__75', 'MS2_all__main_cell_type__50', 'MS2_all__plasma_cell_RPM_ratio', 'MS2_NoCancer__Plasma_Norgen__RPMscaled_prop', 'MS2_NoCancer__Neutrophils__RPMscaled_prop', 'MS2_NoCancer__CD8__RPMscaled_prop', 'MS2_NoCancer__NK_cells__RPMscaled_prop', 'MS2_NoCancer__CD4__RPMscaled_prop', 'MS2_NoCancer__Monocytes__RPMscaled_prop', 'MS2_NoCancer__B_cells__RPMscaled_prop', 'MS2_NoCancer__Erythrocytes__RPMscaled_prop', 'MS2_NoCancer__Thrombocytes__RPMscaled_prop', 'MS2_NoCancer__Basophils__RPMscaled_prop', 'MS2_NoCancer__Eosinophils__RPMscaled_prop', 'MS2_NoCancer__main_cell_type__100', 'MS2_NoCancer__main_cell_type__75', 'MS2_NoCancer__main_cell_type__50', 'MS2_NoCancer__plasma_cell_RPM_ratio', 'MS2_Cancer__Plasma_Norgen__RPMscaled_prop', 'MS2_Cancer__Neutrophils__RPMscaled_prop', 'MS2_Cancer__CD8__RPMscaled_prop', 'MS2_Cancer__NK_cells__RPMscaled_prop', 'MS2_Cancer__CD4__RPMscaled_prop', 'MS2_Cancer__Thrombocytes__RPMscaled_prop', 'MS2_Cancer__Monocytes__RPMscaled_prop', 'MS2_Cancer__B_cells__RPMscaled_prop', 'MS2_Cancer__Basophils__RPMscaled_prop', 'MS2_Cancer__Eosinophils__RPMscaled_prop', 'MS2_Cancer__Erythrocytes__RPMscaled_prop', 'MS2_Cancer__main_cell_type__100', 'MS2_Cancer__main_cell_type__75', 'MS2_Cancer__main_cell_type__50', 'MS2_Cancer__plasma_cell_RPM_ratio']\n",
      "\n",
      ".var sample (first 5 rows):\n",
      "                                  length  g_fraction  a_fraction  t_fraction  \\\n",
      "Sequence                                                                       \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT      32    0.187500    0.250000    0.343750   \n",
      "ATCACAGGGTAGAACCACGGAC                22    0.272727    0.363636    0.090909   \n",
      "TGGAGAGAAAGGCAGTTCCTGT                22    0.363636    0.272727    0.227273   \n",
      "TCCTGACTCCAGGTCCTGTGT                 21    0.238095    0.095238    0.333333   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT            26    0.192308    0.153846    0.269231   \n",
      "\n",
      "                                  c_fraction  gc_fraction  MFE_37  spikein  \\\n",
      "Sequence                                                                     \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT    0.218750     0.406250    -3.6    False   \n",
      "ATCACAGGGTAGAACCACGGAC              0.272727     0.545455    -1.9    False   \n",
      "TGGAGAGAAAGGCAGTTCCTGT              0.136364     0.500000    -1.3    False   \n",
      "TCCTGACTCCAGGTCCTGTGT               0.333333     0.571429    -4.1    False   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT          0.384615     0.576923    -2.0    False   \n",
      "\n",
      "                                  qiaseq_spikein  hbdx_spikein  ...  \\\n",
      "Sequence                                                        ...   \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT           False         False  ...   \n",
      "ATCACAGGGTAGAACCACGGAC                     False         False  ...   \n",
      "TGGAGAGAAAGGCAGTTCCTGT                     False         False  ...   \n",
      "TCCTGACTCCAGGTCCTGTGT                      False         False  ...   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT                 False         False  ...   \n",
      "\n",
      "                                  MS2_Cancer__Thrombocytes__RPMscaled_prop  \\\n",
      "Sequence                                                                     \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT                                  0.000000   \n",
      "ATCACAGGGTAGAACCACGGAC                                            0.019660   \n",
      "TGGAGAGAAAGGCAGTTCCTGT                                            0.015625   \n",
      "TCCTGACTCCAGGTCCTGTGT                                             0.011633   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT                                             NaN   \n",
      "\n",
      "                                  MS2_Cancer__Monocytes__RPMscaled_prop  \\\n",
      "Sequence                                                                  \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT                               0.023641   \n",
      "ATCACAGGGTAGAACCACGGAC                                         0.400515   \n",
      "TGGAGAGAAAGGCAGTTCCTGT                                         0.048490   \n",
      "TCCTGACTCCAGGTCCTGTGT                                          0.438336   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT                                          NaN   \n",
      "\n",
      "                                  MS2_Cancer__B_cells__RPMscaled_prop  \\\n",
      "Sequence                                                                \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT                             0.113767   \n",
      "ATCACAGGGTAGAACCACGGAC                                       0.002137   \n",
      "TGGAGAGAAAGGCAGTTCCTGT                                       0.011537   \n",
      "TCCTGACTCCAGGTCCTGTGT                                        0.018014   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT                                        NaN   \n",
      "\n",
      "                                  MS2_Cancer__Basophils__RPMscaled_prop  \\\n",
      "Sequence                                                                  \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT                                0.00000   \n",
      "ATCACAGGGTAGAACCACGGAC                                          0.00000   \n",
      "TGGAGAGAAAGGCAGTTCCTGT                                          0.00019   \n",
      "TCCTGACTCCAGGTCCTGTGT                                           0.00000   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT                                          NaN   \n",
      "\n",
      "                                  MS2_Cancer__Eosinophils__RPMscaled_prop  \\\n",
      "Sequence                                                                    \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT                                 0.010914   \n",
      "ATCACAGGGTAGAACCACGGAC                                           0.023192   \n",
      "TGGAGAGAAAGGCAGTTCCTGT                                           0.023623   \n",
      "TCCTGACTCCAGGTCCTGTGT                                            0.011595   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT                                            NaN   \n",
      "\n",
      "                                  MS2_Cancer__Erythrocytes__RPMscaled_prop  \\\n",
      "Sequence                                                                     \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT                                  0.000000   \n",
      "ATCACAGGGTAGAACCACGGAC                                            0.215778   \n",
      "TGGAGAGAAAGGCAGTTCCTGT                                            0.271417   \n",
      "TCCTGACTCCAGGTCCTGTGT                                             0.047299   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT                                             NaN   \n",
      "\n",
      "                                  MS2_Cancer__main_cell_type__100  \\\n",
      "Sequence                                                            \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT                        ambiguous   \n",
      "ATCACAGGGTAGAACCACGGAC                                  ambiguous   \n",
      "TGGAGAGAAAGGCAGTTCCTGT                                  ambiguous   \n",
      "TCCTGACTCCAGGTCCTGTGT                                   ambiguous   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT                                    NaN   \n",
      "\n",
      "                                  MS2_Cancer__main_cell_type__75  \\\n",
      "Sequence                                                           \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT                       ambiguous   \n",
      "ATCACAGGGTAGAACCACGGAC                                 ambiguous   \n",
      "TGGAGAGAAAGGCAGTTCCTGT                                 ambiguous   \n",
      "TCCTGACTCCAGGTCCTGTGT                                  ambiguous   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT                                   NaN   \n",
      "\n",
      "                                  MS2_Cancer__main_cell_type__50  \\\n",
      "Sequence                                                           \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT                       ambiguous   \n",
      "ATCACAGGGTAGAACCACGGAC                                 ambiguous   \n",
      "TGGAGAGAAAGGCAGTTCCTGT                                 ambiguous   \n",
      "TCCTGACTCCAGGTCCTGTGT                                  ambiguous   \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT                                   NaN   \n",
      "\n",
      "                                  MS2_Cancer__plasma_cell_RPM_ratio  \n",
      "Sequence                                                             \n",
      "GTGCATGATCTCAAGTTTTCAATCTGAGACCT                           0.000000  \n",
      "ATCACAGGGTAGAACCACGGAC                                     0.000000  \n",
      "TGGAGAGAAAGGCAGTTCCTGT                                     0.281941  \n",
      "TCCTGACTCCAGGTCCTGTGT                                      0.187188  \n",
      "CGTGGTCTCCCAACCCTTGTACCAGT                                      NaN  \n",
      "\n",
      "[5 rows x 186 columns]\n",
      "\n",
      "Fraction of zeros in X: 66.23%\n",
      "Min value in X: 0.0\n",
      "Max value in X: 18.805193502774333\n",
      "None\n",
      "------------------    ------------------  ------------------    -------------------  -----------------  ------------------    ------------------\n",
      "Shape: (180, 432958)\n",
      "X dtype: float64\n",
      "Example values from X: [0.98159156 0.         0.         0.         0.         0.\n",
      " 0.         0.         0.         0.16585449]\n",
      "Layers: []\n",
      "\n",
      ".obs columns: ['total_counts_before_preprocessing', 'total_counts_after_preprocessing', 'total_count_ratio__after_to_before', 'QC_mapping_ratio_bacterial', 'QC_mapping_ratio_viral', 'QC_mapping_ratio_miRNA', 'QC_mapping_ratio_sncRNA', 'Bases', 'Bytes', 'Avg_spot_length', 'BioProject', 'BioSample', 'Experiment', 'SRA_study', 'Sample_name', 'Consent', 'Center_name', 'Organism', 'source_name', 'Participant_number', 'sex', 'age', 'vaccine', 'naat_result', 'timepoint', 'Assay_type', 'Sequencer', 'Sample_type', 'Lab_library_layout', 'Lab_library_selection', 'Lab_library_source', 'Lab_RNA_extraction_protocol', 'Lab_Blocking_State', 'Lab_library_preparation_kit', 'Release_date', 'create_date', 'version', 'total_count']\n",
      "\n",
      ".obs sample (first 5 rows):\n",
      "             total_counts_before_preprocessing  \\\n",
      "smpID                                            \n",
      "SRR24049386                            8351068   \n",
      "SRR24049387                            7921597   \n",
      "SRR24049388                            8299244   \n",
      "SRR24049389                            8186467   \n",
      "SRR24049390                            8671619   \n",
      "\n",
      "             total_counts_after_preprocessing  \\\n",
      "smpID                                           \n",
      "SRR24049386                           8351068   \n",
      "SRR24049387                           7921597   \n",
      "SRR24049388                           8299244   \n",
      "SRR24049389                           8186467   \n",
      "SRR24049390                           8671619   \n",
      "\n",
      "             total_count_ratio__after_to_before  QC_mapping_ratio_bacterial  \\\n",
      "smpID                                                                         \n",
      "SRR24049386                                 1.0                      0.0016   \n",
      "SRR24049387                                 1.0                      0.0013   \n",
      "SRR24049388                                 1.0                      0.0053   \n",
      "SRR24049389                                 1.0                      0.0013   \n",
      "SRR24049390                                 1.0                      0.0015   \n",
      "\n",
      "             QC_mapping_ratio_viral  QC_mapping_ratio_miRNA  \\\n",
      "smpID                                                         \n",
      "SRR24049386                  0.0001                  0.7152   \n",
      "SRR24049387                  0.0001                  0.7347   \n",
      "SRR24049388                  0.0000                  0.5816   \n",
      "SRR24049389                  0.0000                  0.6815   \n",
      "SRR24049390                  0.0000                  0.6393   \n",
      "\n",
      "             QC_mapping_ratio_sncRNA      Bases     Bytes  Avg_spot_length  \\\n",
      "smpID                                                                        \n",
      "SRR24049386                   0.2115  180904905  90231278               21   \n",
      "SRR24049387                   0.2040  172191993  86251440               21   \n",
      "SRR24049388                   0.3206  180833229  91123036               21   \n",
      "SRR24049389                   0.2383  185807806  93750390               22   \n",
      "SRR24049390                   0.2780  191928281  96340055               22   \n",
      "\n",
      "             ... Lab_library_layout Lab_library_selection Lab_library_source  \\\n",
      "smpID        ...                                                               \n",
      "SRR24049386  ...             single    size fractionation     transcriptomic   \n",
      "SRR24049387  ...             single    size fractionation     transcriptomic   \n",
      "SRR24049388  ...             single    size fractionation     transcriptomic   \n",
      "SRR24049389  ...             single    size fractionation     transcriptomic   \n",
      "SRR24049390  ...             single    size fractionation     transcriptomic   \n",
      "\n",
      "            Lab_RNA_extraction_protocol Lab_Blocking_State  \\\n",
      "smpID                                                        \n",
      "SRR24049386       PAXgene blood RNA kit        no blocking   \n",
      "SRR24049387       PAXgene blood RNA kit        no blocking   \n",
      "SRR24049388       PAXgene blood RNA kit        no blocking   \n",
      "SRR24049389       PAXgene blood RNA kit        no blocking   \n",
      "SRR24049390       PAXgene blood RNA kit        no blocking   \n",
      "\n",
      "                Lab_library_preparation_kit          Release_date  \\\n",
      "smpID                                                               \n",
      "SRR24049386  NEBNext Small RNA Library Prep  2024-03-04T00:00:00Z   \n",
      "SRR24049387  NEBNext Small RNA Library Prep  2024-03-04T00:00:00Z   \n",
      "SRR24049388  NEBNext Small RNA Library Prep  2024-03-04T00:00:00Z   \n",
      "SRR24049389  NEBNext Small RNA Library Prep  2024-03-04T00:00:00Z   \n",
      "SRR24049390  NEBNext Small RNA Library Prep  2024-03-04T00:00:00Z   \n",
      "\n",
      "                      create_date version  total_count  \n",
      "smpID                                                   \n",
      "SRR24049386  2023-04-03T14:07:00Z       1      8208137  \n",
      "SRR24049387  2023-04-03T14:07:00Z       1      7784613  \n",
      "SRR24049388  2023-04-03T14:07:00Z       1      8057838  \n",
      "SRR24049389  2023-04-03T14:07:00Z       1      8044754  \n",
      "SRR24049390  2023-04-03T14:07:00Z       1      8533235  \n",
      "\n",
      "[5 rows x 38 columns]\n",
      "\n",
      ".var columns: ['length', 'g_fraction', 'a_fraction', 't_fraction', 'c_fraction', 'gc_fraction', 'MFE_37', 'spikein', 'qiaseq_spikein', 'hbdx_spikein', 'hbdx_spikein_prefix_length', 'hbdx_spikein_suffix_length', 'hbdx_spikein_affix_filter', 'five_prime_adapter_length', 'five_prime_adapter_filter', 'adapter_mapping_filter', 'kmer_five_prime_adapter', 'kmer_three_prime_adapter', 'kmer_filter', 'housekeeping', 'TE_annotation', 'bacterial', 'viral', 'blocking_deseq2R_log2FoldChange', 'blocking_deseq2R_padj', 'reduced_by_blocking', 'blocking_oligo__duplex_energy', 'intended_blocking_target', 'chr', 'num_genomic_loci_maps', 'cluster_gene_name', 'tumor_prog_markers', 'tumor_prog_markers_mutualinfo', 'tumor_prog_markers_meantrend', 'tcga_LC_DESeq2_bool', 'tcga_LC_DESeq2', 'tcga_LC_DESeq2_log2FC', 'tcga_LC_DESeq2_padj', 'tcga_LUSC_DESeq2_bool', 'tcga_LUAD_DESeq2_bool', 'tcga_normal__highest_tissue', 'tcga_normal__highest_mean', 'tcga_normal__TSI', 'tcga_cancer__LUAD_LUSC_DE', 'tcga_cancer__DE_cancers_log2FC', 'tcga_cancer__nr_nonLC_DE_cancers', 'Sozzi', 'Montani', 'HBDx_Masterplate', 'HBDx_IP', 'JAMA', 'General_Literature', 'single_class_annotation', 'single_name_annotation', 'small_RNA_class_annotation', 'subclass_name', 'subclass_name_new_binning', 'subclass_name_new_binning_pos', 'subclass_type', 'pseudo_class', 'mitochondrial', 'ambiguity_resolved', 'hico', 'mms', 'reference', 'precursor_name_full', 'ref_miR_seq', 'miRNA_ref_iso', 'miRNA_hico', 'miRNA_seed', 'miRNA_names', 'tRNA_ref_iso', 'MirGeneDB_ID', 'PTM_count', 'PTM_NGS_mean', 'protein_count', 'protein_affinity_median', 'protein_coverage_perc_median', 'RBCs_RPM_mean', 'CD56+_RPM_mean', 'Neutrophile_RPM_mean', 'CD4+_RPM_mean', 'CD19+_RPM_mean', 'Platelets_RPM_mean', 'Eosinophile_RPM_mean', 'CD8+_RPM_mean', 'CD14+_RPM_mean', 'Basophile_RPM_mean', 'RBCs_RPM_std', 'CD56+_RPM_std', 'Neutrophile_RPM_std', 'CD4+_RPM_std', 'CD19+_RPM_std', 'Platelets_RPM_std', 'Eosinophile_RPM_std', 'CD8+_RPM_std', 'CD14+_RPM_std', 'Basophile_RPM_std', 'RBCs_RPM_prop', 'CD56+_RPM_prop', 'Neutrophile_RPM_prop', 'CD4+_RPM_prop', 'CD19+_RPM_prop', 'Platelets_RPM_prop', 'Eosinophile_RPM_prop', 'CD8+_RPM_prop', 'CD14+_RPM_prop', 'Basophile_RPM_prop', 'Platelets_RPMscaled_mean', 'CD14+_RPMscaled_mean', 'CD56+_RPMscaled_mean', 'CD8+_RPMscaled_mean', 'RBCs_RPMscaled_mean', 'Neutrophile_RPMscaled_mean', 'CD19+_RPMscaled_mean', 'CD4+_RPMscaled_mean', 'Eosinophile_RPMscaled_mean', 'Basophile_RPMscaled_mean', 'Platelets_RPMscaled_std', 'CD14+_RPMscaled_std', 'CD56+_RPMscaled_std', 'CD8+_RPMscaled_std', 'RBCs_RPMscaled_std', 'Neutrophile_RPMscaled_std', 'CD19+_RPMscaled_std', 'CD4+_RPMscaled_std', 'Eosinophile_RPMscaled_std', 'Basophile_RPMscaled_std', 'Platelets_RPMscaled_prop', 'CD14+_RPMscaled_prop', 'CD56+_RPMscaled_prop', 'CD8+_RPMscaled_prop', 'RBCs_RPMscaled_prop', 'Neutrophile_RPMscaled_prop', 'CD19+_RPMscaled_prop', 'CD4+_RPMscaled_prop', 'Eosinophile_RPMscaled_prop', 'Basophile_RPMscaled_prop', 'main_cell_type__100', 'main_cell_type__75', 'main_cell_type__50', 'MS2_all__Plasma_Norgen__RPMscaled_prop', 'MS2_all__Neutrophils__RPMscaled_prop', 'MS2_all__CD8__RPMscaled_prop', 'MS2_all__NK_cells__RPMscaled_prop', 'MS2_all__CD4__RPMscaled_prop', 'MS2_all__Monocytes__RPMscaled_prop', 'MS2_all__B_cells__RPMscaled_prop', 'MS2_all__Erythrocytes__RPMscaled_prop', 'MS2_all__Thrombocytes__RPMscaled_prop', 'MS2_all__Basophils__RPMscaled_prop', 'MS2_all__Eosinophils__RPMscaled_prop', 'MS2_all__main_cell_type__100', 'MS2_all__main_cell_type__75', 'MS2_all__main_cell_type__50', 'MS2_all__plasma_cell_RPM_ratio', 'MS2_NoCancer__Plasma_Norgen__RPMscaled_prop', 'MS2_NoCancer__Neutrophils__RPMscaled_prop', 'MS2_NoCancer__CD8__RPMscaled_prop', 'MS2_NoCancer__NK_cells__RPMscaled_prop', 'MS2_NoCancer__CD4__RPMscaled_prop', 'MS2_NoCancer__Monocytes__RPMscaled_prop', 'MS2_NoCancer__B_cells__RPMscaled_prop', 'MS2_NoCancer__Erythrocytes__RPMscaled_prop', 'MS2_NoCancer__Thrombocytes__RPMscaled_prop', 'MS2_NoCancer__Basophils__RPMscaled_prop', 'MS2_NoCancer__Eosinophils__RPMscaled_prop', 'MS2_NoCancer__main_cell_type__100', 'MS2_NoCancer__main_cell_type__75', 'MS2_NoCancer__main_cell_type__50', 'MS2_NoCancer__plasma_cell_RPM_ratio', 'MS2_Cancer__Plasma_Norgen__RPMscaled_prop', 'MS2_Cancer__Neutrophils__RPMscaled_prop', 'MS2_Cancer__CD8__RPMscaled_prop', 'MS2_Cancer__NK_cells__RPMscaled_prop', 'MS2_Cancer__CD4__RPMscaled_prop', 'MS2_Cancer__Thrombocytes__RPMscaled_prop', 'MS2_Cancer__Monocytes__RPMscaled_prop', 'MS2_Cancer__B_cells__RPMscaled_prop', 'MS2_Cancer__Basophils__RPMscaled_prop', 'MS2_Cancer__Eosinophils__RPMscaled_prop', 'MS2_Cancer__Erythrocytes__RPMscaled_prop', 'MS2_Cancer__main_cell_type__100', 'MS2_Cancer__main_cell_type__75', 'MS2_Cancer__main_cell_type__50', 'MS2_Cancer__plasma_cell_RPM_ratio']\n",
      "\n",
      ".var sample (first 5 rows):\n",
      "                              length  g_fraction  a_fraction  t_fraction  \\\n",
      "Sequence                                                                   \n",
      "GGGAATACCGGGTGCTGTAGGCTG          24    0.458333    0.166667    0.208333   \n",
      "GTGTGGGGTGTAGGTCCGT               19    0.526316    0.052632    0.315789   \n",
      "TCAGGCTCAGTCACCTACCCT             21    0.142857    0.190476    0.238095   \n",
      "AGTGCTGGGATTATAGGAGTGA            22    0.409091    0.272727    0.272727   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT      28    0.392857    0.071429    0.178571   \n",
      "\n",
      "                              c_fraction  gc_fraction  MFE_37  spikein  \\\n",
      "Sequence                                                                 \n",
      "GGGAATACCGGGTGCTGTAGGCTG        0.166667     0.625000    -2.6    False   \n",
      "GTGTGGGGTGTAGGTCCGT             0.105263     0.631579    -1.4    False   \n",
      "TCAGGCTCAGTCACCTACCCT           0.428571     0.571429    -1.3    False   \n",
      "AGTGCTGGGATTATAGGAGTGA          0.045455     0.454545    -1.0    False   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT    0.357143     0.750000    -4.8    False   \n",
      "\n",
      "                              qiaseq_spikein  hbdx_spikein  ...  \\\n",
      "Sequence                                                    ...   \n",
      "GGGAATACCGGGTGCTGTAGGCTG               False         False  ...   \n",
      "GTGTGGGGTGTAGGTCCGT                    False         False  ...   \n",
      "TCAGGCTCAGTCACCTACCCT                  False         False  ...   \n",
      "AGTGCTGGGATTATAGGAGTGA                 False         False  ...   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT           False         False  ...   \n",
      "\n",
      "                              MS2_Cancer__Thrombocytes__RPMscaled_prop  \\\n",
      "Sequence                                                                 \n",
      "GGGAATACCGGGTGCTGTAGGCTG                                           0.0   \n",
      "GTGTGGGGTGTAGGTCCGT                                                0.0   \n",
      "TCAGGCTCAGTCACCTACCCT                                              NaN   \n",
      "AGTGCTGGGATTATAGGAGTGA                                             0.0   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT                                       0.0   \n",
      "\n",
      "                              MS2_Cancer__Monocytes__RPMscaled_prop  \\\n",
      "Sequence                                                              \n",
      "GGGAATACCGGGTGCTGTAGGCTG                                        0.0   \n",
      "GTGTGGGGTGTAGGTCCGT                                             0.0   \n",
      "TCAGGCTCAGTCACCTACCCT                                           NaN   \n",
      "AGTGCTGGGATTATAGGAGTGA                                          0.0   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT                                    0.0   \n",
      "\n",
      "                              MS2_Cancer__B_cells__RPMscaled_prop  \\\n",
      "Sequence                                                            \n",
      "GGGAATACCGGGTGCTGTAGGCTG                                      0.0   \n",
      "GTGTGGGGTGTAGGTCCGT                                           0.0   \n",
      "TCAGGCTCAGTCACCTACCCT                                         NaN   \n",
      "AGTGCTGGGATTATAGGAGTGA                                        0.0   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT                                  0.0   \n",
      "\n",
      "                              MS2_Cancer__Basophils__RPMscaled_prop  \\\n",
      "Sequence                                                              \n",
      "GGGAATACCGGGTGCTGTAGGCTG                                   0.656226   \n",
      "GTGTGGGGTGTAGGTCCGT                                        0.000000   \n",
      "TCAGGCTCAGTCACCTACCCT                                           NaN   \n",
      "AGTGCTGGGATTATAGGAGTGA                                     0.000000   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT                               0.000000   \n",
      "\n",
      "                              MS2_Cancer__Eosinophils__RPMscaled_prop  \\\n",
      "Sequence                                                                \n",
      "GGGAATACCGGGTGCTGTAGGCTG                                          0.0   \n",
      "GTGTGGGGTGTAGGTCCGT                                               0.0   \n",
      "TCAGGCTCAGTCACCTACCCT                                             NaN   \n",
      "AGTGCTGGGATTATAGGAGTGA                                            0.0   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT                                      0.0   \n",
      "\n",
      "                              MS2_Cancer__Erythrocytes__RPMscaled_prop  \\\n",
      "Sequence                                                                 \n",
      "GGGAATACCGGGTGCTGTAGGCTG                                      0.000000   \n",
      "GTGTGGGGTGTAGGTCCGT                                           1.000000   \n",
      "TCAGGCTCAGTCACCTACCCT                                              NaN   \n",
      "AGTGCTGGGATTATAGGAGTGA                                        0.900285   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT                                  1.000000   \n",
      "\n",
      "                              MS2_Cancer__main_cell_type__100  \\\n",
      "Sequence                                                        \n",
      "GGGAATACCGGGTGCTGTAGGCTG                            ambiguous   \n",
      "GTGTGGGGTGTAGGTCCGT                              Erythrocytes   \n",
      "TCAGGCTCAGTCACCTACCCT                                     NaN   \n",
      "AGTGCTGGGATTATAGGAGTGA                              ambiguous   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT                     Erythrocytes   \n",
      "\n",
      "                              MS2_Cancer__main_cell_type__75  \\\n",
      "Sequence                                                       \n",
      "GGGAATACCGGGTGCTGTAGGCTG                           ambiguous   \n",
      "GTGTGGGGTGTAGGTCCGT                             Erythrocytes   \n",
      "TCAGGCTCAGTCACCTACCCT                                    NaN   \n",
      "AGTGCTGGGATTATAGGAGTGA                          Erythrocytes   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT                    Erythrocytes   \n",
      "\n",
      "                              MS2_Cancer__main_cell_type__50  \\\n",
      "Sequence                                                       \n",
      "GGGAATACCGGGTGCTGTAGGCTG                           Basophils   \n",
      "GTGTGGGGTGTAGGTCCGT                             Erythrocytes   \n",
      "TCAGGCTCAGTCACCTACCCT                                    NaN   \n",
      "AGTGCTGGGATTATAGGAGTGA                          Erythrocytes   \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT                    Erythrocytes   \n",
      "\n",
      "                              MS2_Cancer__plasma_cell_RPM_ratio  \n",
      "Sequence                                                         \n",
      "GGGAATACCGGGTGCTGTAGGCTG                                    0.0  \n",
      "GTGTGGGGTGTAGGTCCGT                                         0.0  \n",
      "TCAGGCTCAGTCACCTACCCT                                       NaN  \n",
      "AGTGCTGGGATTATAGGAGTGA                                      0.0  \n",
      "TGGACGCGAGCCGGGCCCTTCCCGTGGT                                0.0  \n",
      "\n",
      "[5 rows x 186 columns]\n",
      "\n",
      "Fraction of zeros in X: 65.46%\n",
      "Min value in X: 0.0\n",
      "Max value in X: 17.147423333682873\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(dp.print_adata_summary(adata_neu))\n",
    "print(\"------------------    ------------------  ------------------    -------------------  -----------------  ------------------    ------------------\")\n",
    "print(dp.print_adata_summary(adata_cov))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16faa6f1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.6622658890935611\n",
      "0.654555137757781\n"
     ]
    }
   ],
   "source": [
    "# Zero fraction\n",
    "print(dp._zero_fraction(adata_neu.X))\n",
    "print(dp._zero_fraction(adata_cov.X))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "21ecfbad",
   "metadata": {},
   "source": [
    "#### batch candidates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9248a261",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.preprocessing import OneHotEncoder\n",
    "from sklearn.linear_model import LinearRegression\n",
    "import scanpy as sc\n",
    "\n",
    "def guess_batch_candidates(adata, exclude=()):\n",
    "    candidates = []\n",
    "    for col in adata.obs.columns:\n",
    "        if col in exclude:\n",
    "            continue\n",
    "        if adata.obs[col].dtype.name == \"category\" or adata.obs[col].dtype == object:\n",
    "            if adata.obs[col].nunique() > 1:\n",
    "                candidates.append(col)\n",
    "    return candidates\n",
    "\n",
    "def _batch_r2_for_pcs(X, expl_var, obs, batch_col):\n",
    "    \"\"\"\n",
    "    PCA for each PC, RÂ² of batch_col\n",
    "    weighted_R2 = variance ratio weighted average\n",
    "    \"\"\"\n",
    "    enc = OneHotEncoder(sparse_output=False, handle_unknown=\"ignore\")\n",
    "    B = enc.fit_transform(obs[[batch_col]])\n",
    "    r2_vec = []\n",
    "    for i in range(X.shape[1]):\n",
    "        y = X[:, i]\n",
    "        mdl = LinearRegression().fit(B, y)\n",
    "        r2_vec.append(mdl.score(B, y))\n",
    "    r2_vec = np.array(r2_vec)\n",
    "    weighted = float((r2_vec * expl_var).sum() / expl_var.sum())\n",
    "    return weighted, r2_vec\n",
    "\n",
    "def rank_batches_within(adata, candidates, base_covars=()):\n",
    "    \"\"\"\n",
    "    Rank batch candidates by weighted_RÂ²\n",
    "    \"\"\"\n",
    "\n",
    "    adata_pca = adata.copy()\n",
    "    sc.pp.highly_variable_genes(adata_pca, flavor=\"seurat_v3\", n_top_genes=2000)\n",
    "    adata_pca = adata_pca[:, adata_pca.var['highly_variable']]\n",
    "    sc.pp.scale(adata_pca, max_value=10)\n",
    "    sc.tl.pca(adata_pca, n_comps=30)\n",
    "\n",
    "    X = adata_pca.obsm['X_pca']\n",
    "    expl_var = adata_pca.uns['pca']['variance_ratio']\n",
    "\n",
    "    rows = []\n",
    "    for col in candidates:\n",
    "        wr2, r2_vec = _batch_r2_for_pcs(X, expl_var, adata.obs, col)\n",
    "        rows.append({'factor': col, 'weighted_partial_R2': wr2, 'n_levels': adata.obs[col].nunique()})\n",
    "    rank_df = pd.DataFrame(rows).sort_values(\"weighted_partial_R2\", ascending=False).reset_index(drop=True)\n",
    "    return rank_df, adata_pca\n",
    "\n",
    "def add_confound_flags(rank_df, adata, bio_label='cell_type', high=0.7):\n",
    "    \"\"\"\n",
    "    add flag column to rank_df based on confounding with bio_label\n",
    "    \"\"\"\n",
    "    df = rank_df.copy()\n",
    "    if bio_label not in adata.obs:\n",
    "        df['confound_flag'] = 'NA'\n",
    "        return df\n",
    "    flags = []\n",
    "    for f in df['factor']:\n",
    "        if f not in adata.obs: \n",
    "            flags.append('NA'); continue\n",
    "        cross = pd.crosstab(adata.obs[f], adata.obs[bio_label])\n",
    "        \n",
    "        chi2 = ((cross - cross.mean(axis=0))**2 / (cross.mean(axis=0)+1e-6)).sum().sum()\n",
    "        norm = chi2 / (len(adata)*len(adata.obs[bio_label].unique()))\n",
    "        if norm > high:\n",
    "            flags.append('high')\n",
    "        else:\n",
    "            flags.append('low')\n",
    "    df['confound_flag'] = flags\n",
    "    return df\n",
    "\n",
    "def deduplicate_batches(adata, ordered, thresh=0.95):\n",
    "    \"\"\"\n",
    "    deduplicate batches if nearly completely confounded\n",
    "    \"\"\"\n",
    "    seen_patterns = []\n",
    "    pruned = []\n",
    "    for f in ordered:\n",
    "        if f not in adata.obs: continue\n",
    "        pattern = tuple(adata.obs[f].astype(str))\n",
    "        dup = False\n",
    "        for p in seen_patterns:\n",
    "            overlap = np.mean([a==b for a,b in zip(pattern, p)])\n",
    "            if overlap >= thresh:\n",
    "                dup = True; break\n",
    "        if not dup:\n",
    "            seen_patterns.append(pattern)\n",
    "            pruned.append(f)\n",
    "    return pruned\n",
    "\n",
    "def select_minimal_batch_set(adata, ordered, base_covars=(), delta=0.01):\n",
    "    \"\"\"\n",
    "    add batch factors from ordered list if partial RÂ² gain >= delta\n",
    "    \"\"\"\n",
    "    sc.pp.highly_variable_genes(adata, flavor=\"seurat_v3\", n_top_genes=2000)\n",
    "    adata_pca = adata[:, adata.var['highly_variable']]\n",
    "    sc.pp.scale(adata_pca, max_value=10)\n",
    "    sc.tl.pca(adata_pca, n_comps=30)\n",
    "    X = adata_pca.obsm['X_pca']\n",
    "    expl_var = adata_pca.uns['pca']['variance_ratio']\n",
    "\n",
    "    selected, gains = [], []\n",
    "    tot = 0.0\n",
    "    for f in ordered:\n",
    "        wr2, _ = _batch_r2_for_pcs(X, expl_var, adata.obs, f)\n",
    "        incr = wr2 - tot\n",
    "        if incr >= delta:\n",
    "            selected.append(f)\n",
    "            gains.append((f, incr))\n",
    "            tot = wr2\n",
    "    return selected, tot, gains\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac3a3956",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "### [neu] within-dataset batch selection ###\n",
      "candidates: ['BioProject', 'BioSample', 'Experiment', 'GEO_Accession_exp', 'SRA_study', 'Sample_name', 'cohort', 'treatment', 'diagnosis', 'create_date']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/scanpy/preprocessing/_highly_variable_genes.py:75: UserWarning: `flavor='seurat_v3'` expects raw count data, but non-integers were found.\n",
      "  warnings.warn(\n",
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/scanpy/preprocessing/_scale.py:316: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              factor  weighted_partial_R2  n_levels confound_flag\n",
      "0          BioSample             1.000000       447            NA\n",
      "1         Experiment             1.000000       447            NA\n",
      "2  GEO_Accession_exp             1.000000       447            NA\n",
      "3        Sample_name             1.000000       447            NA\n",
      "4        create_date             0.389127        14            NA\n",
      "5         BioProject             0.289879         2            NA\n",
      "6          SRA_study             0.289879         2            NA\n",
      "7             cohort             0.289879         2            NA\n",
      "8          treatment             0.289879         2            NA\n",
      "9          diagnosis             0.119703         4            NA\n",
      "after de-dup: ['BioSample', 'Experiment', 'GEO_Accession_exp', 'create_date', 'BioProject', 'SRA_study', 'cohort', 'treatment', 'diagnosis']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/scanpy/preprocessing/_highly_variable_genes.py:75: UserWarning: `flavor='seurat_v3'` expects raw count data, but non-integers were found.\n",
      "  warnings.warn(\n",
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/scanpy/preprocessing/_scale.py:316: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "selected batch set: ['BioSample']  (total weighted partial RÂ² = 1.0 )\n",
      "incremental gains: [('BioSample', 1.0)]\n",
      "\n",
      "### [cov] within-dataset batch selection ###\n",
      "candidates: ['BioSample', 'Experiment', 'Sample_name', 'vaccine', 'naat_result', 'timepoint', 'create_date']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/scanpy/preprocessing/_highly_variable_genes.py:75: UserWarning: `flavor='seurat_v3'` expects raw count data, but non-integers were found.\n",
      "  warnings.warn(\n",
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/scanpy/preprocessing/_scale.py:316: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "        factor  weighted_partial_R2  n_levels confound_flag\n",
      "0    BioSample             1.000000       180            NA\n",
      "1   Experiment             1.000000       180            NA\n",
      "2  Sample_name             1.000000       180            NA\n",
      "3  naat_result             0.115688         3            NA\n",
      "4  create_date             0.035053         3            NA\n",
      "5    timepoint             0.017321         3            NA\n",
      "6      vaccine             0.005348         2            NA\n",
      "after de-dup: ['BioSample', 'Experiment', 'Sample_name', 'naat_result', 'create_date', 'timepoint', 'vaccine']\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/scanpy/preprocessing/_highly_variable_genes.py:75: UserWarning: `flavor='seurat_v3'` expects raw count data, but non-integers were found.\n",
      "  warnings.warn(\n",
      "/mnt/data/venvs/thesis-impute/lib/python3.10/site-packages/scanpy/preprocessing/_scale.py:316: UserWarning: Received a view of an AnnData. Making a copy.\n",
      "  view_to_actual(adata)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "selected batch set: ['BioSample']  (total weighted partial RÂ² = 1.0 )\n",
      "incremental gains: [('BioSample', 1.0)]\n"
     ]
    }
   ],
   "source": [
    "neu = adata_neu\n",
    "cov = adata_cov\n",
    "\n",
    "for name, ad in [('neu', neu), ('cov', cov)]:\n",
    "    print(f\"\\n### [{name}] within-dataset batch selection ###\")\n",
    "    # 1) Explore candidate batches\n",
    "    candidates = guess_batch_candidates(ad, exclude=('sex', 'age', 'study','cell_type','condition','label','donor'))\n",
    "    print(\"candidates:\", candidates)\n",
    "\n",
    "    # 2) Rank candidates by weighted partial RÂ²\n",
    "    rank_df, ad_pca = rank_batches_within(ad, candidates, base_covars=())\n",
    "    rank_df = add_confound_flags(rank_df, ad, bio_label='cell_type', high=0.7)\n",
    "    print(rank_df.head(10))\n",
    "\n",
    "    # 3) Delete duplicates (if any)\n",
    "    ordered = rank_df['factor'].tolist()\n",
    "    pruned = deduplicate_batches(ad, ordered, thresh=0.95)\n",
    "    print(\"after de-dup:\", pruned[:10])\n",
    "\n",
    "    # 4) Select minimal set\n",
    "    selected, tot_w, gains = select_minimal_batch_set(ad, pruned, base_covars=(), delta=0.01)\n",
    "    print(\"selected batch set:\", selected, \" (total weighted partial RÂ² =\", round(tot_w, 3), \")\")\n",
    "    print(\"incremental gains:\", [(f, round(g,3)) for f,g in gains])\n",
    "    \n",
    "    # Visualize top factors and selected\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d8365d6",
   "metadata": {},
   "source": [
    "### Denormalize back to counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "bef4679b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Zero fraction (counts): 0.6622658890935611\n",
      "Row-sum corr vs totals: 0.9999999980500865\n",
      "Median abs diff: 314.0\n",
      "Min/Max counts: 0 3857841\n",
      "Zero fraction (counts): 0.654555137757781\n",
      "Row-sum corr vs totals: 0.9999997938787484\n",
      "Median abs diff: 6923.5\n",
      "Min/Max counts: 0 1599031\n"
     ]
    }
   ],
   "source": [
    "for adata in [adata_neu, adata_cov]:\n",
    "    dp.invert_log2cpm_to_counts(\n",
    "        adata,\n",
    "        totals_col=\"total_counts_after_preprocessing\",\n",
    "        out_layer=\"counts\",\n",
    "        clip_negative=True\n",
    "    )\n",
    "\n",
    "adata_neu.write(\"/mnt/data/thesis_project/data/processed/raw_count/GSE169569_raw_counts.h5ad\")\n",
    "adata_cov.write(\"/mnt/data/thesis_project/data/processed/raw_count/GSE228841_raw_counts.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bbff775b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KeysView(Layers with keys: counts, log2_1p_CPM_original)\n",
      "KeysView(Layers with keys: counts, log2_1p_CPM_original)\n"
     ]
    }
   ],
   "source": [
    "IN_PATH_neu = \"../data/processed/raw_count/GSE169569_raw_counts.h5ad\"\n",
    "IN_PATH_cov = \"../data/processed/raw_count/GSE228841_raw_counts.h5ad\"\n",
    "\n",
    "adata_neu = sc.read_h5ad(IN_PATH_neu)\n",
    "adata_cov = sc.read_h5ad(IN_PATH_cov)\n",
    "\n",
    "print(adata_neu.layers.keys())\n",
    "print(adata_cov.layers.keys())"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "thesis-impute",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
